app/grandchallenge/reader_studies/forms.py

```diff
@@ -478,41 +480,174 @@ def clean_ground_truth(self):
-                else:
-                    # Use a ModelChoiceField here, as InterfaceFormField would
-                    # provide an upload wodget, but we do not want to add new
-                    # images/files here, but rather assign existing values to
-                    # the proper display sets.
-                    self.fields[slug] = ModelChoiceField(
-                        queryset=ComponentInterfaceValue.objects.filter(
-                            id__in=civ["values"]
-                        ),
-                        initial=val,
-                        required=False,
-                    )
-
-            self.fields["order"] = IntegerField(initial=instance.order)
+        self.instance = instance
+        self.reader_study = reader_study
+        self.user = user
+        for slug, values in reader_study.values_for_interfaces.items():
+            current_value = None
+            if instance:
+                current_value = instance.values.filter(
+                    interface__slug=slug
+                ).first()
+            interface = ComponentInterface.objects.get(slug=slug)
+            if interface.is_image_kind:
+                self.fields[slug] = self._get_image_field(
+                    interface, values, current_value
+                )
+            elif interface.requires_file:
+                self.fields[slug] = self._get_file_field(
+                    interface, values, current_value
+                )
+            else:
+                self.fields[slug] = self._get_json_field(
+                    interface, current_value
+                )
+
+        order = (
+            instance.order if instance else reader_study.next_display_set_order
+        )
+        self.fields["order"] = IntegerField(initial=order)
+
+    def _get_default_field(self, interface, current_value):
+        return InterfaceFormField(
+            instance=interface,
+            initial=current_value.value if current_value else None,
+            required=False,
+            user=self.user,
+        ).field
+
+    def _get_json_field(self, interface, current_value):
+        return self._get_default_field(interface, current_value)
+
+    def _get_image_field(self, interface, values, current_value):
+        return ModelChoiceField(
+            queryset=ComponentInterfaceValue.objects.filter(id__in=values),
+            initial=current_value,
+            required=False,
+            widget=SelectUploadWidget(
+                attrs={
+                    "reader_study_slug": self.reader_study.slug,
+                    "display_set_pk": self.instance.pk,
+                    "interface_pk": interface.pk,
+                }
+            ),
+        )
+
+    def _get_file_field(self, interface, values, current_value):
+        return self._get_image_field(interface, values, current_value)
+
+
+class DisplaySetUpdateForm(DisplaySetBaseForm):
+    def __init__(self, *args, instance, user, **kwargs):
```

Comment:
This `__init__` method has a different signature to the base, dangerous. Why overload it at all?